Add ARD overpass notebook + supplementary data

Eric-git-999 opened the pull request:
Supplementary data for ARD_overpass_predictor notebook in dea-notebooks/Frequently_used_code/

### Proposed changes
Supplementary data for ARD_overpass_predictor notebook in dea-notebooks/Frequently_used_code/

### Checklist (replace `[ ]` with `[x]` to check off)
- [x] Notebook created using the [DEA-notebooks template](https://github.com/GeoscienceAustralia/dea-notebooks/tree/develop)
- [x] Remove any unused Python packages from `Load packages`
- [x] Remove any unused/empty code cells
- [x] Remove any guidance cells (e.g. `General advice`)
- [x] Ensure that all code cells follow the [PEP8 standard](https://www.python.org/dev/peps/pep-0008/) for code. The `jupyterlab_code_formatter` tool can be used to format code cells to a consistent style: select each code cell, then click `Edit` and then one of the `Apply X Formatter` options (`YAPF` or `Black` are recommended).
- [x] Include relevant tags in the final notebook cell (refer to the [DEA Tags Index](https://docs.dea.ga.gov.au/genindex.html), and re-use tags if possible)
- [x] Clear all outputs, run notebook from start to finish, and save the notebook in the state where all cells have been sequentially evaluated
- [x] Test notebook on both the `NCI` and `DEA Sandbox` (flag if not working as part of PR and ask for help to solve if needed)
- [x] If applicable, update the `Notebook currently compatible with the NCI|DEA Sandbox environment only` line below the notebook title to reflect the environments the notebook is compatible with

Eric-git-999 committed:
Add files via upload

Supplementary data for ARD_overpass_predictor notebook in dea-notebooks/Frequently_used_code/

MatthewJA commented:
Hi Eric, thanks for your PR. I can't seem to find the ARD overpass notebook in question? Could you please provide a link? Thanks.

Eric-git-999 commented:
Sorry it’s coming I swear!

I had to edit the path in the notebook to specify this example file and was just testing that in the Sandbox. It all works now so will be submitting the notebook shortly.

Thanks

From: Matthew Alger <notifications@github.com>
Sent: Monday, 7 December 2020 10:47 AM
To: GeoscienceAustralia/dea-notebooks <dea-notebooks@noreply.github.com>
Cc: Hay Eric <Eric.Hay@ga.gov.au>; Author <author@noreply.github.com>
Subject: Re: [GeoscienceAustralia/dea-notebooks] Add input dataset for ARD overpass notebook (#736)


Hi Eric, thanks for your PR. I can't seem to find the ARD overpass notebook in question?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/GeoscienceAustralia/dea-notebooks/pull/736#issuecomment-739586939>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANJD6OJ3IYKT4L7LUTKPZTLSTQJYHANCNFSM4UPT24NA>.

Geoscience Australia Disclaimer: This e-mail (and files transmitted with it) is intended only for the person or entity to which it is addressed. If you are not the intended recipient, then you have received this e-mail by mistake and any use, dissemination, forwarding, printing or copying of this e-mail and its file attachments is prohibited. The security of emails transmitted cannot be guaranteed; by forwarding or replying to this email, you acknowledge and accept these risks.

MatthewJA commented:
Great, looking forward to it :) Just add it to this PR and we can look at the notebook + data at the same time.

Eric-git-999 committed:
Add files via upload

Eric-git-999 commented:
Hey I have added the notebook!

Eric-git-999 committed:
Add files via upload

Eric-git-999 commented:
Tidied up the fist MD cell with a link to DEA Sandbox, and the DEA image

MatthewJA commented on Frequently_used_code/ARD_overpass_predictor.ipynb:
```diff
@@ -0,0 +1,1677 @@
+{
```
<p>I think you can assume that the user will change their input file if they want to analyse somewhere else. Instead of telling the user to change the file, explain how the file is formatted (which you've done below anyway) and they can make that change if they want. So I reckon remove the "Caution" line.</p><p> </p><p>The input file is <code>csv</code> now right, not <code>xlsx</code> ?</p><p> </p><p>I don't quite understand this line, as I think I should be able to run the notebook without making nchanges:</p><blockquote>Make changes to the notebook, following the <strong>Steps</strong> in bold</blockquote><p> </p><p>Load packages shouldn't be a subsection of description, it should be in a getting started section - check the template and make sure you're matching how it is formatted and structured.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

MatthewJA replied:
<p>Why are the secondary overpasses not already in datetime format? I think you can get <code>pd.read_csv</code>  to force them all into datetime format automatically.</p><p> </p><p>What is a secondary overpass and why would I want to have one in my input? Could you please add a little explanation?</p><p> </p><p>Don't use <code>os.chdir</code> as it has a tendency to make the rest of the notebook harder to understand and may break existing scripts.  Read the file using a relative path instead. You also can't assume that this notebook is in <code>jovyan/</code> (e.g. I am testing it in a different place!) so try something like <code>../Supplementary_data/ARD_overpass_predictor/overpass_input.csv</code>  instead.</p><p> </p><p>Input file looks good!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

MatthewJA replied:
<p>If there's no reason not to use the more accurate timestamps, let's use those.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

MatthewJA replied:
<p>What's the significance of 20? I'm a bit confused as to what this is doing. I think it's finding the next 20 times Landsat will be overhead? Please make this a bit clearer.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

MatthewJA replied:
<p>Again, why 32?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

MatthewJA replied:
<p>The Landsat, Sentinel 2A, and Sentinel 2B code are all very similar. Can you use a dictionary and make this into a loop, so that the code isn't repeated? Alternatively you could write a function that operated on one satellite, and then apply it for all three.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

MatthewJA replied:
<p>If this is evident from the input file, then we should be able to automatically extract it from the input file. Please do that instead of having to edit the notebook if at all possible.</p><p> </p><p>Also, <code>overpass_input.csv</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

MatthewJA replied:
<p>I'm not sure what the datestep means. Is it a meaningful value? If not, maybe we could just give the rows an index?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

MatthewJA replied:
<p>Maybe we can output all of the field sites that were in the input? That would make it easier.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

MatthewJA replied:
<p>to_csv rather than to_excel</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

MatthewJA replied:
<p>It's a good idea to have the output filename at the top of the notebook along with other analysis parameters like the input name.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

MatthewJA commented:
This is a really useful notebook @EricHay, and it'll be a great addition to Frequently_used_code! A few comments so far that I've posted on ReviewNB, and a few I'll post here. My main request is that you add more documentation. This looks like a useful tool, but `dea-notebooks` is an entry point for much of the DEA environment, so we want all notebooks to be well-explained and detailed so that even beginners can understand them. Take a look at some of the other Frequently_used_code notebooks - that's the level of documentation I'd like to see! Could you please add more markdown cells that explain what you're doing and why so that people without a strong background in the topic can understand what's going on? There's also a lot of repetition going on between the satellites and locations, which could be eliminated by some loops or functions. That'd make the notebook much more easy to edit and work with, and so it'd be much more useful! Thanks :)

Eric-git-999 commented:
Great thanks Matthew. I sure can fix up the documentation and code. I did this a while back as a learning exercise and it is a pretty convoluted process!

robbibt commented:
Hey @EricHay, notebook looks awesome! To echo some of what Matthew wrote above, I think the key changes that are needed are:

1. Extra markdown documentation and explanation from a really basic beginner level, explaining both what each section does, but also why it is needed. A lot of users of `dea-notebooks` have never dealt with complex code before, so we try and walk them through everything pretty slowly. This is particularly the case for the `Frequently_used_code` directory which is kind of like a "library" where users can jump in and copy and paste examples of code into their own analyses - we want them to have as good an idea as possible about what each section does so they can re-use and re-purpose the code themselves later on.

2. Wherever possible, it would be great if all bits of code that require user input be moved up the very top in an "analysis parameters" section ([see example here](https://docs.dea.ga.gov.au/notebooks/Frequently_used_code/Analyse_multiple_polygons.html)). We typically find that users somewhat blindly run through most of the notebook body without paying close attention to things that need to be changed later in the code, so having all the configurable bits in one single up-front section means they only need to focus on changing one bit rather than having to be on the look out for making changes all the way through in order to get valid results. 

Happy to help out with any of this stuff! It's looking great through, and will be an excellent addition to the repo. 🚀

Eric-git-999 commented:
Thanks @robbibt, sorry i'm a bit flat out today with meetings but will get to polishing this off soon. Agree on both points 👍 and thanks for the feedback!

MatthewJA commented:
Take your time, no rush :)

Also feel free to Slack me (or Robbi probably?) if you need any help figuring out how to document it.

Eric-git-999 committed:
Add files via upload

Eric-git-999 committed:
Add files via upload

Eric-git-999 commented:
<p>Ok I have done some major tidying. I have simplified things A LOT! It is actually kinda fun to re-visit old notebooks and fix them up.</p><p> </p><p>The notebook is now pretty much automatic. No more specifying which sites to order by etc, and will merge using pd.concat instead of pd.merge which was messy.  You can specify an output directory / file name at the top of the notebook. I have also simplified the input file to use 3 sites as an example, and added hashed out lines where you can add extra sites. </p><p> </p><p>The only comment I couldn't really address was about repitition around the looping for different satellites. I am not sure how to "loop loops" :p </p><p>I combined these all into one cell and tried to explain it a bit better.</p><p>Let me know if I should change anything further!</p>

Eric-git-999 committed:
Add files via upload

Eric-git-999 commented:
Sorry I forgot to remove some text referring to the old Process. Should be good now!

Eric-git-999 committed:
Add files via upload

robbibt commented on Frequently_used_code/ARD_overpass_predictor.ipynb:
```diff
@@ -0,0 +1,864 @@
+{
```
<p>These commented out bits make the code a little harder to follow:</p><p> </p><pre class="ql-syntax" spellcheck="false">#Sentinel_2A = Sentinel_2A + datetime.timedelta(hours=10) #convert to local time (Aus eastern standard time) = utc + 10 hours
#Sentinel_2A&nbsp;### to AEDT, add 11h not 10 ###
</pre><p> </p><p>I think it might be better to simplify the code by removing them from here and instead make it really clear in the <code>Combine dataset</code>  bit that the times are in UTC (?) (and possibly give an example of how to convert the columns in the final table there). </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

robbibt replied:
<p>Same as above: can we remove the commented out lines here? Anyone who wants to look at the data can always edit and add them in themselves</p><p> </p><pre class="ql-syntax" spellcheck="false">#S2B_combined
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

robbibt replied:
<p>This table looks really nice and useful! The "Site" title for the first column is a tiny bit confusing (perhaps it should be "Overpass"?) but that's a super minor thing and probably not worth updating.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GeoscienceAustralia/dea-notebooks/pull/736/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

robbibt commented:
Hey @EricHay , I just posted a few very minor comments above, it's looking great! Thanks for all your work in updating this, I think it should be much easier to use now.

Eric-git-999 commented:
Thanks @robbibt. Easy enough fixes, I agree on all points. This has been a bit of a backburner project, and I didn't even notice the "Site" column in the final table! That can definitely be dropped. I will do another quick tidy, and comment again once done.

Eric-git-999 committed:
Add files via upload

Eric-git-999 commented:
Ok, tidied up the unnecessary commented-out code in cells, updated the final table with "Overpass" as the index, and added optional code in the Combine Dataset section to add / alter time zones, with UTC to AEST as an example (+10h). I think this is straight-forward enough for an average person with some Python knowledge to utilise now :)

robbibt reviewed (approved):
Whoops, did I forget to accept this? I think this looks much nicer and easier to use/follow, thanks for putting in these changes! 

When you're happy to merge, select the "squash and merge" option below :)
